### PR TITLE
SIMS #174 - Auth e2e test

### PR DESCRIPTION
**Pull request by andrewsignori-aot**
Enabled E2E test to test authentication scenarios as shown below, following the recommendations here: https://docs.nestjs.com/fundamentals/testing#end-to-end-testing

In general, the E2E tests are executed in two steps:

1. Get a valid token from Keyclock;
2. Try to authenticate to a fake API (but with real functionality) controller (AuthTestController) 

image

Key observations:

- To allow us to test the E2E authentication cycle, a dummy user (student_e2e_test) was added to the Keycloak DEV instance.
- This user is used to retrieve the token from Keycloak and authenticate the API endpoints.
- The AuthTestController controller was created to be the target for all E2E tests related to the authentication. This controller is not part of the final API (not available on DEV/TEST/PROD) and it is being used only for the E2E tests (see comments on the file auth-test.controller.ts). 
- The Keycloak endpoints are retrieved from the discovery endpoint ".well-known/openid-configuration".
- To run the tests locally the ENV variable E2E_TEST_STUDENT_PASSWORD need to be defined in a local .env file, to avoid add the Keyclock dummy user password in the source code.
- It is not part of this PR the enablement of E2E on the CI/CD. We need to do this for UnitTest/E2E Test/Lint, etc.
- Along the process, some changes affected the Unit Tests already in place and some changes were needed to keep them working.
image

### CODE REVIEW

File: sources/packages/api/src/services/config/config.service.ts [hunk 1]
```diff
@@ -9,14 +9,20 @@ export class ConfigService {
         url: process.env.KEYCLOAK_AUTH_URL,
         realm: process.env.KEYCLOAK_REALM,
         clientId: process.env.KEYCLOAK_CLIENT,
-        issuerUrl: this.getIssuerEndpoint(),
+        OpenIdConfigurationUrl: this.getOpenIdConfigurationUrl(),
+      },
+      e2eTest: {
```
Comment (fwpushan) [inline sources/packages/api/src/services/config/config.service.ts:14]: Do we need to provide this information here? I suggest, somewhere in test or testHelper

**Review by fwpushan — CHANGES_REQUESTED** [review]
Overall structure and KeyCloak Auth related changes are good but my only concerns is there are few test related config inside main application sources.

File: sources/packages/api/src/types/config.ts [hunk 2]
```diff
@@ -1,10 +1,20 @@
 export interface IConfig {
   auth: IAuthConfig;
+  e2eTest: Ie2eTest;
```
Comment (fwpushan) [inline sources/packages/api/src/types/config.ts:3]: Can we move this into test or testHelpers?

File: sources/packages/api/src/types/config.ts [hunk 3]
```diff
@@ -1,10 +1,20 @@
...
 
 export interface IAuthConfig {
   url: string;
   realm: string;
   clientId: string;
-  issuerUrl: string;
+  OpenIdConfigurationUrl: string;
+}
+
+export interface Ie2eTest {
```
Comment (fwpushan) [inline sources/packages/api/src/types/config.ts:13]: Can we move all test related config into test or testHelpers ?

**Review by fwpushan — CHANGES_REQUESTED** [review]

File: sources/packages/api/src/types/config.ts [hunk 4]
```diff
@@ -6,5 +6,5 @@ export interface IAuthConfig {
   url: string;
   realm: string;
   clientId: string;
-  issuerUrl: string;
+  OpenIdConfigurationUrl: string;
```
Comment (fwpushan) [inline sources/packages/api/src/types/config.ts:9]: why this property name starts with Capital letter?

**Review by fwpushan — COMMENTED** [review]
Please check there is only a minor comment.

File: sources/packages/api/src/auth/roles.enum.ts [hunk 5]
```diff
@@ -3,4 +3,5 @@
  */
 export enum Role {
   Student = "StudentRole",
+  DummyRole = "DummyRole",
```
[the remark this answers is not in the document]
Comment (andrewsignori-aot) [reply sources/packages/api/src/auth/roles.enum.ts]: It is removed!

File: sources/packages/api/src/services/config/config.service.ts (hunk 1 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/services/config/config.service.ts]: I removed it from the main source as we talked.

File: sources/packages/api/src/types/config.ts (hunk 4 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/types/config.ts:9]: Because it is wrong :D
I am gonna change to the proper case.

Comment (andrewsignori-aot) [reply sources/packages/api/src/types/config.ts]: Fixed!

**Review by fwpushan — APPROVED** [review]

**Comment by andrewsignori-aot** [comment]
Went through a peer review chat with @2011sandeepraj and got the approval to execute the merge.

File: sources/packages/api/src/types/config.ts (hunk 3 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/types/config.ts]: Done!

File: sources/packages/api/src/types/config.ts (hunk 2 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/types/config.ts]: Done!

**Pull request merged**